docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,80 @@
+
+## Motivation
+
+At present, the handling of VPN connectivity lacks consistency for users. There are instances when the VPN goes into No Signal due to a variety of connectivity-related issues, such as the network connection failing or the server location becoming unavailable. In these situations, users are often left without a clear explanation of what went wrong or guidance on how to troubleshoot and resolve the issue. They may encounter a background error message or lose signal without further information.
+
+The objective with the Connection Manager is to address these issues by running probes on all potential causes and presenting the user with an appropriate modal detailing the root cause of their problem. This approach aims to enhance the user's understanding of the issue and provide them with actionable steps to rectify it.
+
+To avoid further complicating the already complex logic of the controller, this document proposes encapsulating this new functionality within the Connection Manager. This object will interact with both the controller and connection health, serving as a dedicated entity for identifying connection problems and correctly communicating status changes to the UI, ultimately enhancing the user experience.
+
+This proposal describes Connection Manager behavior during activation. Subsequent proposal(s) will address the behavior of the Connection Manager during the lifespan of a connection.
+
+## Proposed Solution
+
+We will introduce a new object, the Connection Manager, which will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionManager::State`. The Connection Manager will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful. 
+
+In the event of a probe failure, the Connection Manager will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted. Once the issue has been resolved, the cycle of probes commences anew. If there are no ongoing probes, the Connection Manager will enter an Idle state.
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Manager.
+
```

Comment:
> We can leverage this logic to recover from a probe failure within the Connection Manager.

Could this interaction between ConnectionManager and ConnectionHealth be described in more detail? For example, if ConnectionHealth's probe indicates NoSignal, how does ConnectionManager use that information to identify the underlying cause?

My understanding is that when ConnectionHealth signals Unstable or NoSignal to ConnectionManager, the latter will run its probes to determine the underyling cause and then communicate that to the front end.

Reply:
> My understanding is that when ConnectionHealth signals Unstable or NoSignal to ConnectionManager, the latter will run its probes to determine the underyling cause and then communicate that to the front end.

Connection Diagnostics will not do anything if the Connection Health stability mode is `Unstable`. We only initiate the diagnostics check if stability is NoSignal